src/nanotron/trainer.py

```diff
@@ -246,24 +251,77 @@ def post_train_step(self):
     def post_training(self):
         pass
 
+    def _print_training_plan(self):
+        stages_info = "".join(
+            f"[Stage {stage.name}] start from step {stage.start_training_step} \n" for stage in self.config.data_stages
+        )
+        full_log_message = f"[Training Plan] There are {len(self.config.data_stages)} training stages \n{stages_info}"
+        log_rank(full_log_message, logger=logger, level=logging.INFO, rank=0)
+
+    def _update_dataloader_based_on_training_stages(self, dataloaders: List[DataLoader]):
+        assert len(dataloaders) > 0, "No dataloaders provided"
+        assert len(dataloaders) == len(
+            self.config.data_stages
+        ), "Number of dataloaders should match the number of dataset stages"
+
+        def clear_dataloader_from_memory(dataloader: DataLoader, stage_name: str):
+            import gc
+
+            log_rank(
+                f"[Training Stage: {stage_name}] Clearing the previous training stage's dataloader and datasets from memory",
+                logger=logger,
+                level=logging.INFO,
+            )
+
+            # NOTE: Clear the internal state of the dataloader
+            del dataloader.dataset
+            del dataloader.sampler
+            del dataloader.batch_sampler
+
+            gc.collect()
```

Comment:
Is the gc.collect necessary? Doesnt it happen automatically? Did you try with and without and notice a big diff?

Reply:
Because it directly frees unused objects and does not require waiting for the garbage collector to be scheduled... It would be nice to directly call it :)